### PR TITLE
VueForms module

### PR DESCRIPTION
**Pull request by jptissot**
Implement a Forms module that allows easy integration with Vuetify / VeeValidate. 

- Integrated client / server side validation
- Easily create content items from the form without a workflow
- dynamic server side validation

TODO: 
- Cypress tests

### CODE REVIEW

File: src/Modules/StatCan.OrchardCore.AjaxForms/Controllers/AjaxFormController.cs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                return Json(new { validationError = true, html = formHtml });
+            }
+
+            if (!string.IsNullOrEmpty(script?.OnSubmitted?.Text))
+            {
+                _scriptingManager.EvaluateJs(script.OnSubmitted.Text, scriptingProvider);
+            }
+
+            // _workflow manager is null if workflow feature is not enabled
+            if (_workflowManager != null)
```
Comment (sebastienros) [inline src/Modules/StatCan.OrchardCore.AjaxForms/Controllers/AjaxFormController.cs:105]: This won't be null if the feature is not enabled, the ctor would just fail.

Comment (jptissot) [reply src/Modules/StatCan.OrchardCore.AjaxForms/Controllers/AjaxFormController.cs:105]: This did worked, as the ctor takes a optional parameter. I was not sure how to do this otherwise.

**Review by jptissot — COMMENTED** [review]
Delete file

File: .idea/.idea.StatCan.OrchardCore/.idea/.gitignore [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿# Default ignored files
```
Comment (jptissot) [inline .idea/.idea.StatCan.OrchardCore/.idea/.gitignore:1]: Delete File

File: .idea/.idea.StatCan.OrchardCore/.idea/encodings.xml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jptissot) [inline .idea/.idea.StatCan.OrchardCore/.idea/encodings.xml:1]: Delete

File: .idea/.idea.StatCan.OrchardCore/.idea/indexLayout.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jptissot) [inline .idea/.idea.StatCan.OrchardCore/.idea/indexLayout.xml:1]: Delete

File: .idea/.idea.StatCan.OrchardCore/.idea/misc.xml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jptissot) [inline .idea/.idea.StatCan.OrchardCore/.idea/misc.xml:1]: Delete

File: .idea/.idea.StatCan.OrchardCore/riderModule.iml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jptissot) [inline .idea/.idea.StatCan.OrchardCore/riderModule.iml:1]: Delete

File: .idea/.idea.StatCan.OrchardCore/.idea/vcs.xml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jptissot) [inline .idea/.idea.StatCan.OrchardCore/.idea/vcs.xml:1]: Delete

File: src/Build/Dependencies.props [hunk 8]
```diff
@@ -8,7 +8,7 @@
 
   <ItemGroup>
     <PackageManagement Include="NetEscapades.AspNetCore.SecurityHeaders" Version="0.11.0" />
-    <PackageManagement Include="HtmlSanitizer" Version="4.0.217" />
+    <PackageManagement Include="HtmlSanitizer" Version="5.0.319" />
```
Comment (jptissot) [inline src/Build/Dependencies.props:11]: Do this in a seperate PR

File: src/Build/StatCan.OrchardCore.Commons.targets [hunk 9]
```diff
@@ -40,10 +40,6 @@
       <UnmanagedPackageReference Remove="@(UnmanagedPackageReference)" Condition=" '%(UnmanagedPackageReference.IsImplicitlyDefined)' == 'true' " />
       <UnmanagedPackageReference Remove="@(UnmanagedPackageReference)" Condition=" $([System.String]::Copy('%(Identity)').StartsWith('System.')) " />
     </ItemGroup>
-    <Warning Condition=" '@(UnmanagedPackageReference)' != '' and '%(Identity)' != 'Microsoft.AspNetCore.App' "
```
Comment (jptissot) [inline src/Build/StatCan.OrchardCore.Commons.targets:43]: Do this in a seperate PR

File: src/Modules/StatCan.OrchardCore.AjaxForms/ControllerExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewEngines;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+
+namespace StatCan.OrchardCore.AjaxForms
+{
+    // todo: move this into it's own assembly.
+    public static class ControllerExtensions
+    {
+        public static async Task<string> RenderViewAsync<TModel>(this Controller controller, string viewName, TModel model, bool partial = false)
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.AjaxForms/ControllerExtensions.cs:14]: Might be useful to move this in a seperate assembly

File: src/Modules/StatCan.OrchardCore.AjaxForms/Models/AjaxFormPart.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class AjaxForm : ContentPart
+    {
+        public BooleanField Enabled { get; set; }
+
+        public TextField SuccessMessage { get; set; }
+        public TextField ErrorMessage { get; set; }
+
+        // todo: feature that helps debug forms in the admin ui
+        // public BooleanField DebugValues { get; set; }
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.AjaxForms/Models/AjaxFormPart.cs:14]: Clean up file

File: src/Modules/StatCan.OrchardCore.AjaxForms/Models/AjaxFormScripts.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
+using OrchardCore.ContentFields.Fields;
+using OrchardCore.ContentManagement;
+
+namespace StatCan.OrchardCore.AjaxForms.Models
+{
+    public class AjaxFormScripts: ContentPart
+    {
+        /// <summary>
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.AjaxForms/Models/AjaxFormScripts.cs:8]: Add a Initial values script

File: src/Modules/StatCan.OrchardCore.AjaxForms/Models/FormRequiredValidation.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public class FormRequiredValidation : ContentPart
+    {
+        // validation
+        public BooleanField Required { get; set; }
+        public TextField RequiredText { get; set; }
+    }
+    public class FormInputValidation: ContentPart
+    {
+        //public NumericField MinLength { get; set; }
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.AjaxForms/Models/FormRequiredValidation.cs:14]: Finish this implementation. Maybe leave the validation for a seperate PR

File: src/Modules/StatCan.OrchardCore.AjaxForms/Validation/FormClientModelValidationProvider.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
+//using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.AjaxForms/Validation/FormClientModelValidationProvider.cs:1]: Remove File

File: src/Modules/StatCan.OrchardCore.AjaxForms/Validation/FormModelValidationProvider.cs [hunk 15]
```diff
@@ -0,0 +1,24 @@
+//using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.AjaxForms/Validation/FormModelValidationProvider.cs:1]: Remove File

File: src/Modules/StatCan.OrchardCore.ContentFields/Views/TextField-CodeMirrorJS.Edit.cshtml [hunk 16]
```diff
@@ -0,0 +1,32 @@
+@model OrchardCore.ContentFields.ViewModels.EditTextFieldViewModel
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.ContentFields/Views/TextField-CodeMirrorJS.Edit.cshtml:1]: Move this to another PR

File: test/cypress/integration/03-AjaxForms/01-setup-forms.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
+/// <reference types="Cypress" />
```
Comment (jptissot) [inline test/cypress/integration/03-AjaxForms/01-setup-forms.js:1]: Fix the tests

File: src/Modules/StatCan.OrchardCore.VueForms/Views/Content-VueForm.cshtml [hunk 18]
```diff
@@ -0,0 +1,14 @@
+@await DisplayAsync(Model.Content.Named("FlowPart"))
+
+@*Todo: localresource*@
+<style asp-src="https://cdn.jsdelivr.net/npm/vuetify@2.3.10/dist/vuetify.min.css" debug-src="https://cdn.jsdelivr.net/npm/vuetify@2.3.10/dist/vuetify.css"></style>
```
[the remark this answers is not in the document]
Comment (jptissot) [reply src/Modules/StatCan.OrchardCore.VueForms/Views/Content-VueForm.cshtml]: Not entirely possible without implementing a webpack build for this module.

File: src/Modules/StatCan.OrchardCore.VueForms/Models/VueFormPart.cs [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+namespace StatCan.OrchardCore.VueForms.Models
+{
+    public class VueForm : ContentPart
+    {
+        public BooleanField Enabled { get; set; }
+
+        public TextField SuccessMessage { get; set; }
+        public TextField ErrorMessage { get; set; }
+
+        // todo: feature that helps debug forms in the admin ui
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.VueForms/Models/VueFormPart.cs:13]: Remove commented out code

File: docs/en/reference/modules/VueForms.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+#### ClientInit script
+
+The ClientInit script is executed client side prior to instanciating the VueJS app. This script has access to the client's global scope variables.
+
+This was added to support hooking into some global options for VeeValidate.
+
+For example, setting the [VeeValidate.setInteractionMode('passive')](https://logaretm.github.io/vee-validate/guide/interaction-and-ux.html) option 
+or the [Localization](https://logaretm.github.io/vee-validate/guide/localization.html) option.
+
+It is also possible to modify the vuetify [global-config](https://vuetifyjs.com/en/customization/global-config/) or [presets](https://vuetifyjs.com/en/customization/presets/).
```
Comment (jptissot) [inline docs/en/reference/modules/VueForms.md:39]: Show how this is possible

File: src/Modules/StatCan.OrchardCore.GCCollab/README.md [hunk 21]
```diff
@@ -3,4 +3,6 @@
 Orchard Core module providing GCCollab Authentication.
 
 ## Configuration 
-First, enable the GCCollab module. Then go to Security->Authentication->GCCollab and enter the `client id`, `client secret` and `redirect url`. Last, enable the `User Registration` feature then go to Settings->Registration and select `AllowRegistration`.
+First, enable the GCCollab module.
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.GCCollab/README.md:6]: Move this file to the docs and it should not be modified by this PR

File: src/Modules/StatCan.OrchardCore.VueForms/Assets/vue-forms.js [hunk 22]
```diff
@@ -0,0 +1,183 @@
+"use strict";
+
+// register VeeValidate components globally
+Vue.component('validation-provider', VeeValidate.ValidationProvider);
+Vue.component('validation-observer', VeeValidate.ValidationObserver); 
+
+// include default english and french translations.
+VeeValidate.localize({
+  en:{
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.VueForms/Assets/vue-forms.js:9]: Maybe there is a cleaner way to do this. Not 100% sure.

File: src/Modules/StatCan.OrchardCore.VueForms/Assets/vue-forms.js [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+      "regex": "Le champ {_field_} est invalide",
+      "required": "Le champ {_field_} est obligatoire",
+      "required_if": "Le champ {_field_} est obligatoire lorsque {target} possède cette valeur",
+      "size": "Le champ {_field_} doit avoir un poids inférieur à {size}KB"
+    }
+  },
+});
+
+var vuetify = new Vuetify();
+
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.VueForms/Assets/vue-forms.js:72]: We should probably tell VeeValidate which language is currently being used around this line.

File: src/Modules/StatCan.OrchardCore.VueForms/Controllers/VueFormController.cs [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+
+            var formPart = form.As<VueForm>();
+
+            // Verify if the form is enabled
+            if (!formPart.Enabled.Value)
+            {
+                return NotFound();
+            }
+
+
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.VueForms/Controllers/VueFormController.cs:76]: Remove white line

File: src/Modules/StatCan.OrchardCore.VueForms/Scripting/VueFormMethodsProvider.cs [hunk 25]
```diff
@@ -0,0 +1,48 @@
+using System;
+using System.Collections.Generic;
+using AngleSharp.Common;
+using OrchardCore.ContentManagement;
+using OrchardCore.Scripting;
+
+namespace StatCan.OrchardCore.VueForms.Scripting
+{
+    public class VueFormMethodsProvider : IGlobalMethodProvider
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.VueForms/Scripting/VueFormMethodsProvider.cs:9]: Add a note why this does not need to be added to the Startup.cs file

File: src/Modules/StatCan.OrchardCore.VueForms/Views/Content-VueForm.cshtml [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            @await DisplayAsync(widgetContent)
+          }
+          <v-form>
+      </validation-observer>
+    </v-app>
+  </div>
+} else
+{
+  @await DisplayAsync(Model.Content.Named("VueForm-DisabledHtml"))
```
Comment (jptissot) [inline src/Modules/StatCan.OrchardCore.VueForms/Views/Content-VueForm.cshtml:58]: Test this flow.
